lib/db/queries.ts

```diff
@@ -0,0 +1,159 @@
+    return sql`${books.title_tsv} @@ to_tsquery('english', ${tsQuery})`;
+  }
+  return undefined;
+};
+
+const imageFilter = () => {
+  return and(
+    not(isNull(books.image_url)),
+    sql`${books.image_url} != ${EMPTY_IMAGE_URL}`
+  );
+};
+
+const isbnFilter = (isbn?: string) => {
+  if (isbn) {
+    const isbnArray = isbn.split(',').map((id) => id.trim());
+    return sql`books.isbn IN (${sql.join(
+      isbnArray.map((id) => sql`${id}`),
+      sql`, `
+    )})`;
+  }
+  return undefined;
+};
+
+export async function fetchBooksWithPagination(searchParams: SearchParams) {
+  let requestedPage = Math.max(1, Number(searchParams?.page) || 1);
+
+  const filters = [
+    yearFilter(searchParams.yr),
+    ratingFilter(searchParams.rtg),
+    languageFilter(searchParams.lng),
+    pageFilter(searchParams.pgs),
+    imageFilter(),
+    searchFilter(searchParams.search),
+    isbnFilter(searchParams.isbn),
+  ].filter(Boolean);
+
+  const whereClause = filters.length > 0 ? and(...filters) : undefined;
+  const offset = (requestedPage - 1) * ITEMS_PER_PAGE;
+
+  const paginatedBooks = await db
+    .select({
+      id: books.id,
+      title: books.title,
+      image_url: books.image_url,
+      thumbhash: books.thumbhash,
+    })
+    .from(books)
+    .where(whereClause)
+    .orderBy(books.id)
+    .limit(ITEMS_PER_PAGE)
+    .offset(offset);
+
+  return paginatedBooks;
+}
+
+export async function estimateTotalBooks(searchParams: SearchParams) {
+  const filters = [
+    yearFilter(searchParams.yr),
+    ratingFilter(searchParams.rtg),
+    languageFilter(searchParams.lng),
+    pageFilter(searchParams.pgs),
+    imageFilter(),
+    searchFilter(searchParams.search),
+    isbnFilter(searchParams.isbn),
+  ].filter(Boolean);
+
+  const whereClause = filters.length > 0 ? and(...filters) : undefined;
+
+  // Use explain to get an estimate
+  const explainResult = await db.execute(sql`
+    EXPLAIN (FORMAT JSON)
+    SELECT id FROM books
+    ${whereClause ? sql`WHERE ${whereClause}` : sql``}
+  `);
+
+  const planRows = (explainResult.rows[0] as any)['QUERY PLAN'][0]['Plan'][
+    'Plan Rows'
+  ];
+  return planRows;
+}
+
+export async function fetchBookById(id: string) {
+  let result = await db
+    .select({
+      id: books.id,
+      isbn: books.isbn,
+      title: books.title,
+      publication_year: books.publication_year,
+      publisher: books.publisher,
+      image_url: books.image_url,
+      description: books.description,
+      num_pages: books.num_pages,
+      language_code: books.language_code,
+      text_reviews_count: books.text_reviews_count,
+      ratings_count: books.ratings_count,
+      average_rating: books.average_rating,
+      series: books.series,
+      createdAt: books.createdAt,
+      authors: sql<string[]>`array_agg(${authors.name})`,
+      thumbhash: books.thumbhash,
+    })
+    .from(books)
+    .leftJoin(bookToAuthor, eq(books.id, bookToAuthor.bookId))
```

Comment:
You could omit indexes I've mentioned below if you'd query from `bookToAuthor` table first with joins on `books.id` and `authors.id`